### PR TITLE
[Feat] Refactor comments CRUD

### PR DESCRIPTION
**Pull request by aadamw**
## Description
- fixed https://github.com/typehero/typehero/issues/1095
- Moved comments/replies logic into it's separate hooks
- `CommentsInput` no longer have 2 input values states (external and internal)
- Deleting last comment on 2 (or other page different than 1), not automatically goes back to previous page

## Related Issue
Closes https://github.com/typehero/typehero/issues/1095

## Motivation and Context
This came-up as work trying to fix https://github.com/typehero/typehero/issues/1095. So I decided to refactor it a little bit.
Comments components are kinda mess right now, and full of prop drilling. As a next step I'd love to fix composition over there, so it would be easier to maintain and extend if needed in the future

### CODE REVIEW

File: apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      // If the last comment on the page was deleted, we need to go back a page
+      const newPage = data?.comments.length === 1 ? meta.page - 1 : meta.page;
+      changePage(newPage);
+      queryClient.invalidateQueries({
+        queryKey: getQueryKey({ sort: meta.sort.value, page: newPage }),
+      });
+    }
+  };
+
+  const addComment = async (text: string) => {
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx:108]: should these be using `useCallback` looks like youre passing some of these down to components

Comment (aadamw) [reply apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx:108]: IMO, not necessary since all actions will be re-invalidating queries, which will trigger refetches, and deps will change in those `useCallbacks`. 
I haven't noticed any performance issues so far, so let's not try to fix it prematurely

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx [hunk 2]
```diff
@@ -14,65 +12,38 @@ import { singleFieldSchema, type SingleFieldSchema } from '~/utils/zodSingleStri
...
-
-      setIsSubmitting(true);
-
-      await onSubmit();
+      await onSubmit(text);
     } catch (e) {
       console.error(e);
     } finally {
-      setIsSubmitting(false);
+      form.reset();
```
Comment (ayrock-dev) [inline apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:40]: praise: nice work I like the form.reset() here, :+1: web standards

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx [hunk 3]
```diff
@@ -14,65 +12,38 @@ import { singleFieldSchema, type SingleFieldSchema } from '~/utils/zodSingleStri
...
-    }
-    setIsSubmitting(true);
-    await onSubmit();
-    setIsSubmitting(false);
-  };
-
-  const submitComment = async () => {
+  const submitComment = async ({ text }: { text: string }) => {
     try {
-      if (!session?.user) {
```
Comment (ayrock-dev) [inline apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:55]: (non-blocking) question: do we still capture the logged-in requirement somewhere else?

File: apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+interface DefaultCommentsProps {
+  rootId: number;
+  type: CommentRoot;
+}
+
+interface UseCommentsProps extends DefaultCommentsProps {
+  initialPage?: number;
+}
+
+export function useComments({ type, rootId, initialPage }: UseCommentsProps) {
```
Comment (ayrock-dev) [inline apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx:39]: praise: thanks for extracting these into a hook, reads very clearly

File: apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    } catch (e) {
+      toast({
+        ...commentErrors.unexpected,
+        variant: 'destructive',
+      });
+    } finally {
+      // If the last comment on the page was deleted, we need to go back a page
+      const newPage = data?.comments.length === 1 ? meta.page - 1 : meta.page;
+      changePage(newPage);
+      queryClient.invalidateQueries({
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx:102]: this should only happen when successful?

File: apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+      } else if (res === 'unauthorized') {
+        toast(commentErrors.unauthorized);
+      }
+    } catch (e) {
+      toast({
+        ...commentErrors.unauthorized,
+        variant: 'destructive',
+      });
+    } finally {
+      queryClient.invalidateQueries({
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx:150]: should this happen only when successful?

File: apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+  type: CommentRoot;
+}
+
+interface UseCommentsProps extends DefaultCommentsProps {
+  initialPage?: number;
+}
+
+export function useComments({ type, rootId, initialPage }: UseCommentsProps) {
+  const queryClient = useQueryClient();
+  const [meta, updateMeta] = useReducer(
```
Comment (ayrock-dev) [inline apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx:41]: thought: this file reads great, but I did have to scratch my head a bit to figure out what `meta` is and where we create that. For a minute I thought I was looking at [meta](https://tanstack.com/query/v4/docs/react/guides/query-functions#queryfunctioncontext) from @tanstack/query.

**Review by ayrock-dev — COMMENTED** [review]
LGTM, I have not pulled this locally to test the "gotchas"

a couple of the edge cases look like-
 * delete a comment and try going to the share url on your clipboard
 * delete a reply and try going to the share url on your clipboard
 * delete a handful of comments and then try expanding the "Read more" to see the next page, make sure deleted comments don't show up and we didn't skip any comments

**Review by ayrock-dev — APPROVED** [review]

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx (hunk 3 above)
Comment (aadamw) [reply apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:55]: Yes, it was duplicated logic, that's why I removed it

File: apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        });
+      }
+    } catch (e) {
+      toast({
+        ...commentErrors.unexpected,
+        variant: 'destructive',
+      });
+    } finally {
+      // If the last comment on the page was deleted, we need to go back a page
+      const newPage = data?.comments.length === 1 ? meta.page - 1 : meta.page;
```
Comment (zaCKoZAck0) [inline apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx:100]: It Should not move to prev page on error scenarios as well. If I'm clearly understanding the login it may.

**Review by zaCKoZAck0 — APPROVED** [review]

**Review by bautistaaa — APPROVED** [review]
